windows/src/global/delphi/debug/Keyman.System.DebugLogManager.pas

```diff
@@ -0,0 +1,284 @@
+  public
+    constructor Create(AOwner: HWND); reintroduce;
+    destructor Destroy; override;
+
+  end;
+
+function GetDebugManager(AOwner: HWND): TDebugLogManager;
+
+implementation
+
+uses
+  Accctrl,
+  AclApi,
+  ErrorControlledRegistry,
+  ErrLogPath,
+  RegistryKeys,
+  Unicode;
+
+var
+  FDebugManager: TDebugLogManager = nil;
+  FTerminating: Boolean = False;
+
+  FShouldDebug: Boolean = False;
+  FShouldDebugLastTick: Cardinal = 0;
+
+const
+  // GUID that identifies our trace session.
+  SessionGuid: TGUID = '{FADEA67E-0EE9-452B-AF04-22E342D1227A}';
+
+function GetDebugManager(AOwner: HWND): TDebugLogManager;
+begin
+  Assert(not FTerminating);
+
+  if ShouldDebug then
+  begin
+    if not Assigned(FDebugManager) then
+      FDebugManager := TDebugLogManager.Create(AOwner);
+    Result := FDebugManager;
+  end
+  else
+    Result := nil;
+end;
+
+{ TDebugManager }
+
+const LOGSESSION_NAME: string = 'Keyman Event Trace Session';
+
+function StringBufferSize(const s: string): Integer;
+begin
+  Result := (Length(s) + 1) * sizeof(WCHAR);
+end;
+
+constructor TDebugLogManager.Create(AOwner: HWND);
+var
+  BufferSize: ULONG;
+  status: ULONG;
+begin
+  FOwner := AOwner;
+
+  FindFirstLogFileName;
+  StartNewLogFile;
+
+  // Allocate memory for the session properties. The memory must
+  // be large enough to include the log file name and session name,
+  // which get appended to the end of the session properties structure.
+
+  BufferSize := sizeof(EVENT_TRACE_PROPERTIES) + StringBufferSize(FDebugLogFileName) + StringBufferSize(LOGSESSION_NAME);
+  pSessionProperties := PEVENT_TRACE_PROPERTIES(AllocMem(BufferSize));
+
+  // Set the session properties. You only append the log file name
+  // to the properties structure; the StartTrace function appends
+  // the session name for you.
+
+  pSessionProperties.Wnode.BufferSize := BufferSize;
+  pSessionProperties.Wnode.Flags := WNODE_FLAG_TRACED_GUID;
+  pSessionProperties.Wnode.ClientContext := 1; //QPC clock resolution
+  pSessionProperties.Wnode.Guid := SessionGuid;
+  pSessionProperties.LogFileMode := EVENT_TRACE_FILE_MODE_SEQUENTIAL;
+  pSessionProperties.MaximumFileSize := 1;  // 1 MB
+  pSessionProperties.LoggerNameOffset := sizeof(EVENT_TRACE_PROPERTIES);
+  pSessionProperties.LogFileNameOffset := sizeof(EVENT_TRACE_PROPERTIES) + sizeof(LOGSESSION_NAME);
+  StrPCopy(PWideChar(PByte(pSessionProperties) + pSessionProperties.LogFileNameOffset), FDebugLogFileName);
+
+  status := StartTrace(@FSessionHandle, PWideChar(LOGSESSION_NAME), pSessionProperties^);
+  if ERROR_ALREADY_EXISTS = status then
+  begin
+    // The trace was already started, perhaps Keyman did not close down cleanly
+    // We'll stop it and restart it
+    status := ControlTraceW(FSessionHandle, PWideChar(LOGSESSION_NAME), pSessionProperties^, EVENT_TRACE_CONTROL_STOP);
+    if ERROR_SUCCESS <> status then
+      OutputDebugString(PChar('ControlTrace failed with '+IntToStr(status)));
```

Comment:
Since there's `ControlTraceW`, `ControlTraceA`, and `ControlTrace`, do you want to specify `ControlTraceW` here?

also in line 216.

Reply:
No, it's only debug messaging, so it'll be clear enough. ControlTraceW is the Unicode version and ControlTraceA is the ANSI version. ControlTrace in Delphi is, these days, mapped to ControlTraceW anyway.